File: src/algorithms/linkedlist/InsertionSortList.java
package algorithms.linkedlist; 

import java.util.*; 
import lib.ListNode; 

public class InsertionSortList {
    /* Function: Use insertion sorting to sort a list. 
     * Solution: The previous part are sorted. Find the position of current
     *     element in the previous part. Be careful when current is larger 
     *     than all elements in the previous part. 
     */ 
    
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) {
            return head; 
        }
        
        ListNode dummy = new ListNode(0); 
        dummy.next = head; 
        
        ListNode done = head, cur = done.next; 
        while (cur != null) {
            ListNode run = dummy; 
            while (run.next != cur && run.next.val < cur.val) {
                run = run.next; 
            }
            
            if (run.next != cur) {
                done.next = cur.next; 
                cur.next = run.next; 
                run.next = cur; 
            } else {
                done = cur; 
            }
            cur = done.next; 
        }
        
        return dummy.next; 
    }
}

File: src/CPP/src/leetcode/InsertionSortList.cpp
/*
 * [Source] https://leetcode.com/problems/insertion-sort-list/
 * [Difficulty]: Medium
 * [Tag]: Linked List
 * [Tag]: Sort
 */

#include <iostream>

using namespace std;

// [Solution]: Keep a list with sorted nodes. When reach the new node, move it into the right place.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/linkedlist/InsertionSortList.java
 */

/* Java solution
public class Solution {
    public ListNode insertionSortList(ListNode head) {
		if(head==null || head.next==null) return head;
		ListNode current=head.next;
		ListNode prev=head;
		while(current!=null){
			ListNode run=new ListNode(0);
			run.next=head;
			boolean get=false;
			while(run.next!=current){
				if(run.next.val>current.val){
					if(run.next==head) head=current;
					prev.next=current.next; // sepearate current
					// insert current
					current.next=run.next;
					run.next=current;
					current=prev.next;
					get=true;
					break;
				}
				run=run.next;
			}
			if(!get){
				current=current.next;
				prev=prev.next;
			}
		}
		return head;
	}
}
 */

int main() {
    Solution sol;

    return 0;
}
